### ISSUE TITLE
1000s of calls to child_process.exec

### ISSUE THREAD

**Issue by kolodny** [google/google-auth-library-nodejs#187]
When a new instance of a google auth instance is created, it looks calls the `getDefaultProjectId` method. This method does caching, however the cache only really works once the first call settles. More aggressive caching can be called for if there is a loop creating hundreds of instances. The following is something I threw together which solves the problem. I'm not really familiar with ts enough yet to create a PR but the patch should look something like this:

<strike>

```diff
diff --git a/ts/lib/auth/googleauth.ts b/ts/lib/auth/googleauth.ts
index deb2b90..77b999b 100644
--- a/ts/lib/auth/googleauth.ts
+++ b/ts/lib/auth/googleauth.ts
@@ -32,6 +32,8 @@ import {JWT} from './jwtclient';
 import {OAuth2Client} from './oauth2client';
 import {UserRefreshClient} from './refreshclient';

+let getDefaultProjectIdPromise = null;
+
 export interface ProjectIdCallback { (err: Error, projectId: string): void }

 export interface CredentialBody {
@@ -126,14 +128,33 @@ export class GoogleAuth {
         this.callback(callback, null, this._cachedProjectId);
       });
     } else {
+      let resolve = null;
+      let reject = null;
       const my_callback = (err: Error, projectId: string) => {
         if (!err && projectId) {
           this._cachedProjectId = projectId;
+          if (resolve) {
+            resolve(projectId);
+          }
+        } else {
+          if (reject) {
+            reject(error);
+          }
         }
         setImmediate(() => {
           this.callback(callback, err, projectId);
         });
       };
+      if (getDefaultProjectIdPromise) {
+        getDefaultProjectIdPromise.then(
+          (projectId) => my_callback(null, projectId),
+          (error) => my_callback(error),
+        );
+        return;
+      }
+      getDefaultProjectIdPromise = new Promise((res, rej) => {
+        resolve = res;
+        reject = rej;
+      });

       // environment variable
       if (this._getProductionProjectId(my_callback)) {
```

</strike>

<sub>Edit: That last bit was an overly complex fix, see the next comment instead</sub>

The following file highlights the calls to exec from this library:

```js
const childProcess = require('child_process');

const oldExec = childProcess.exec;
childProcess.exec = function() {
  console.log(`exec called ${new Error().stack.split('\n')[2].trim()}`);
  return oldExec.apply(this, arguments);
}

const BigTable = require('@google-cloud/bigtable');
const rowCount = 10;

const INSTANCE_ID = process.env.INSTANCE_ID
if (!INSTANCE_ID) {
  return console.error('need to set INSTANCE_ID');
}

(async () => {
  try {
    const bigtable = BigTable();
    const instance = bigtable.instance(INSTANCE_ID);

    const tableName = `write_test${Math.random().toString().slice(2)}`;
    const table = instance.table(tableName);
    const [tableExists] = await table.exists();
    if (!tableExists) {
      console.log(`Creating table ${tableName}`);
      const options = { families: ['cf'] };
      await table.create(options);
    }
    for (let index = 0; index < rowCount; index++) {
      const bigtable = BigTable();
      const instance = bigtable.instance(INSTANCE_ID);
      const table = instance.table(tableName)
      const row = {
        key: `userid_${index}`,
        data: {
          cf: { index },
        },
      };
      table.insert(row, (err) => {
        console.log('wrote row')
        if (err) console.error('error:', error)
      })
    }
  } catch (error) {
    console.error(`caught ${error}`);
  }
})();

```

Which outputs:

```
exec called at GoogleAuth._getSDKDefaultProjectId (~demo/node_modules/google-auth-library/lib/auth/googleauth.js:232:3)
Creating table write_test22080385082919474
exec called at GoogleAuth._getSDKDefaultProjectId (~demo/node_modules/google-auth-library/lib/auth/googleauth.js:232:3)
exec called at GoogleAuth._getSDKDefaultProjectId (~demo/node_modules/google-auth-library/lib/auth/googleauth.js:232:3)
exec called at GoogleAuth._getSDKDefaultProjectId (~demo/node_modules/google-auth-library/lib/auth/googleauth.js:232:3)
exec called at GoogleAuth._getSDKDefaultProjectId (~demo/node_modules/google-auth-library/lib/auth/googleauth.js:232:3)
exec called at GoogleAuth._getSDKDefaultProjectId (~demo/node_modules/google-auth-library/lib/auth/googleauth.js:232:3)
exec called at GoogleAuth._getSDKDefaultProjectId (~demo/node_modules/google-auth-library/lib/auth/googleauth.js:232:3)
exec called at GoogleAuth._getSDKDefaultProjectId (~demo/node_modules/google-auth-library/lib/auth/googleauth.js:232:3)
exec called at GoogleAuth._getSDKDefaultProjectId (~demo/node_modules/google-auth-library/lib/auth/googleauth.js:232:3)
exec called at GoogleAuth._getSDKDefaultProjectId (~demo/node_modules/google-auth-library/lib/auth/googleauth.js:232:3)
exec called at GoogleAuth._getSDKDefaultProjectId (~demo/node_modules/google-auth-library/lib/auth/googleauth.js:232:3)
wrote row
wrote row
wrote row
wrote row
wrote row
wrote row
wrote row
wrote row
wrote row
wrote row
```

**Comment by kolodny** [comment]
Looking into this a bit more, it appears the proper fix would be to make `_cachedProjectId` a top level var and not an property of each class instance, since there's a new instance of the class for each call to `table.insert(...)`. The same could also apply for `_cachedCredential`

**Comment by JustinBeckwith** [comment]
I don't think we want to have a cache that sits outside of the class instance, but #216 will cache the Promise instead of the ProjectId.  `table.insert` should not be creating a new authClient every time - if it is, we should probably open a bug over in that repo:
https://github.com/googleapis/nodejs-bigquery

Assuming the PRs land, this will be released as part of the 1.0 package.  After that, I can help figure out what's going on in bigquery.

**Comment by ofrobots** [comment]
What version of `@google-cloud/bigquery` was used in that example? Relatively recent versions of that module already uses `google-auto-auth` library which caches the auth client. Does the problem still occur with the latest version?

**Comment by JustinBeckwith** [comment]
Landed in #216

**Comment by kolodny** [comment]
@ofrobots You wouldn't be able to replicate since https://github.com/GoogleCloudPlatform/google-cloud-node/pull/2755 got merged and released, since caching the proto object will circumvent the call to exec
